builtin/credential/approle/path_role.go

```diff
@@ -2355,9 +2375,38 @@ func (b *backend) handleRoleSecretIDCommon(ctx context.Context, req *logical.Req
 		return nil, err
 	}
 
+	var numUses int
+	// Check whether or not num_uses is defined, otherwise fallback to secret_id_num_uses
+	if numUsesRaw, ok := data.GetOk("num_uses"); ok {
+		numUses = numUsesRaw.(int)
+		if numUses < 0 {
+			return logical.ErrorResponse("num_uses cannot be negative"), nil
+		}
+
+		// If the role's secret_id_num_uses is higher than the specified num_uses, throw an error rather than implicitly overriding
+		if (numUses == 0 && role.SecretIDNumUses > 0) || (role.SecretIDNumUses > 0 && numUses > role.SecretIDNumUses) {
+			return logical.ErrorResponse("num_uses cannot be higher than the role's secret_id_num_uses"), nil
```

Comment:
> This error message confuses me a little, specifically for the first case. If numUses == 0 && role.SecretIDNumUses > 0 then num_uses isn't higher than the role's secret_id_num_uses, is it? It's lower.

Whenever either num_uses or ttl is set to zero it equals to [unlimited/infinite](https://www.vaultproject.io/api-docs/auth/approle#:~:text=A%20value%20of%20zero%20will%20allow%20unlimited%20uses.). This guard statement's first part ensures that no unlimited custom num_uses can be used whenever the role's secret id configuration uses a limited value. The second part ensures the normal flow, as long as the role's secret id configuration does not use an unlimited num_uses value.
The same applies to the ttl parameter.

> Additionally, wouldn't this get triggered by default, e.g. if num_uses isn't set (because it would default to 0)?

No this wouldn't get triggered by default. The `data#GetOk` method checks if a parameter was specified, if not, it defaults to the Role's Secret ID num_uses. This was done to not break the current implementation. The same applies to the ttl parameter.

Reply:
Ah, I'm following you now, thanks. I do think the comment might be incorrect though, as it states that an error should be thrown if the role's secret_id_num_uses is higher than the specified num_uses, but the code is opposite of that (and I think in this case the code is correct).

Reply:
You are correct. I will port this together while rephrasing. This was already part of the plan as it wasn't sure yet before what the approach would be. 😄